RFCs/0002-pluggable-discovery.md

```diff
@@ -0,0 +1,568 @@
+# Pluggable Discovery protocol (Version 1.0.0)
+
+## Overview
+
+This document describes how the Pluggable Discovery works and how it should integrate with monitors and uploaders.
+
+## Problem
+
+When a sketch is uploaded to an Arduino board the only way for transferring the binary executable to the microcontroller is through the serial port. Currently:
+
+- the subroutines to enumerate the serial ports and to possibly identify a board are “hardcoded” in the IDE/CLI
+- the only way to identify a board is via USB VID/PID identifiers
+- the "serial monitor", to communicate with the board after the upload, is “hardcoded” in the IDE
+
+The current structure does not allow to use different kind of “ports” to communicate with the microcontroller. For example it would be interesting to have the possibility to:
+
+- discover a board (MDNS) and upload via network (OTA)
+
+- discover a board and upload via a communication port that is not the serial port (for example via USB-HID as implemented by Teensy or via any other kind of port)
+
+- allow a third party to implement his own upload protocol / port
+
+The pluggable discovery aims to provide a solution to these problems.
+
+## Constraints
+
+- A communication port may use any kind of **protocol**
+
+- Software support for a new **protocol** must be added to the system as a **plugin**
+
+- Each port must be **enumerated/discovered**
+
+- Each port may provide **properties relative to the port** (USB serial number / MAC address)
+
+- Each port must have an **unique address and protocol pair**
+
+- A single device may expose multiple ports
+
+## Recommended Solution
+
+The proposed solution is to provide a tool that can enumerate and discover the ports for a **single specific protocol**. There will be a tool for each supported protocol so, in the end, we will have a “Serial ports discovery”, a “Network port discovery” and so on.
+
+These tools must be in the form of executables that can be launched as a subprocess using a `platform.txt` command line recipe. They will communicate to the parent process via stdin/stdout, in particular a discovery will accept commands as plain text strings from stdin and will send answers back in JSON format on stdout. Each tool will implement the commands to list and enumerate ports for a specific protocol as specified in this document.
+
+### Pluggable discovery API via stdin/stdout
+
+All the commands listed in this specification must be implemented in the discovery.
+
+After startup, the tool will just stay idle waiting for commands. The available commands are: `HELLO`, `START`, `STOP`, `QUIT`, `LIST` and `START_SYNC`.
```

Comment:
The protocol now defines commands and the corresponding replies, but does not specify:
 - Commands and reply sequencing (i.e. should you wait for a reply before sending a new command? Can a command return multiple replies? How does `START_SYNC` play into this?)
 - Any max response time or timeout
 - Any error responses

Reply:
I've added some more details about message sequencing and timings.